tools/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/ActiveTransactionRegistry.java

```diff
@@ -94,24 +105,59 @@ private ActiveTransactionRegistry() {
 
 		secondaryCache = CacheBuilder.newBuilder()
 				.removalListener((RemovalNotification<UUID, Transaction> notification) -> {
+					logger.debug("secondary cache removal");
 					if (RemovalCause.EXPIRED.equals(notification.getCause())) {
 						final UUID transactionId = notification.getKey();
 						final Transaction entry = notification.getValue();
+						logger.debug("expired transaction to be removed {}", transactionId);
 						synchronized (primaryCache) {
-							if (!entry.hasActiveOperations()) {
-								// no operation active, we can decommission this entry
-								primaryCache.invalidate(transactionId);
-								logger.warn("deregistered expired transaction {}", transactionId);
-							} else {
-								// operation still active. Reinsert in secondary cache.
-								getSecondaryCache().put(transactionId, entry);
+							// no operation active, we can decommission this entry
+							primaryCache.invalidate(transactionId);
+							logger.debug("deregistered expired transaction {}", transactionId);
+							try {
+								logger.debug("try close() invoked on transaction !!!{}", transactionId);
+								entry.close();
+							} catch (Throwable t) {
+								logger.debug("error on close when purging {}", t.getMessage());
 							}
 						}
 					}
 				})
 				.expireAfterAccess(timeout, TimeUnit.SECONDS)
 				.build();
+		cleaupSecondaryCacheScheduler = Executors.newSingleThreadScheduledExecutor((
+
+				Runnable runnable) -> {
+			Thread thread = Executors.defaultThreadFactory().newThread(runnable);
+			thread.setName("rdf4j-cleanup-stn-scheduler");
+			thread.setDaemon(true);
+			return thread;
+		});
+
+		// timeout + 10% to force cleanup
+		cleanupTask = cleaupSecondaryCacheScheduler.schedule(() -> {
+			cleanUpSecondaryCache();
+		}, timeout + timeout / 10, TimeUnit.SECONDS);
+		logger.debug("secondary cache expire time {} seconds", timeout);
+	}
```

Comment:
Normally, cache.cleanUp would automatically trigger whenever something touches the cache (writes to it). I think we originally assumed that this cache would have sufficient throughput for that to happen automatically, without an explicit call. Do I understand correctly that there's cases where that doesn't happen? If so, having this cleanup thread makes sense to me.

Reply:
Since it is touched only when remote explicit transactions are made, but thse are usually far less frequent than regular updates (which is 95% of the cases) it may stay like that indefinitely.